File: ios/SsdidWallet/Feature/Recovery/KeyRotationScreen.swift
import SwiftUI

struct KeyRotationScreen: View {
    @Environment(AppRouter.self) private var router
    @EnvironmentObject private var services: ServiceContainer

    let keyId: String

    enum RotationState {
        case idle
        case preparing
        case rotating
        case success(String)
        case error(String)
    }

    @State private var identity: Identity?
    @State private var state: RotationState = .idle
    @State private var hasPreCommitment = false
    @State private var nextKeyHash: String?

    struct RotationHistoryEntry: Identifiable {
        let id = UUID()
        let oldKeyFragment: String
        let newKeyFragment: String
        let timestamp: String
    }

    @State private var rotationHistory: [RotationHistoryEntry] = []

    private var inProgress: Bool {
        if case .preparing = state { return true }
        if case .rotating = state { return true }
        return false
    }

    var body: some View {
        VStack(spacing: 0) {
            // Header
            HStack(spacing: 4) {
                Button { router.pop() } label: {
                    Image(systemName: "chevron.left")
                        .foregroundStyle(Color.textPrimary)
                        .font(.system(size: 20))
                }
                .padding(.leading, 8)

                Text("Key Rotation")
                    .font(.ssdidHeadline)
                    .foregroundStyle(Color.textPrimary)

                Spacer()
            }
            .padding(.vertical, 12)
            .padding(.trailing, 20)

            ScrollView {
                LazyVStack(spacing: 10) {
                    // Current key info
                    Text("CURRENT KEY")
                        .font(.ssdidCaption)
                        .foregroundStyle(Color.textTertiary)
                        .frame(maxWidth: .infinity, alignment: .leading)

                    if let id = identity {
                        VStack(alignment: .leading, spacing: 0) {
                            HStack {
                                Text("KEY ID")
                                    .font(.system(size: 11))
                                    .foregroundStyle(Color.textTertiary)
                                Spacer()
                                AlgorithmBadge(
                                    name: id.algorithm.rawValue.replacingOccurrences(of: "_", with: "-"),
                                    isPostQuantum: id.algorithm.isPostQuantum
                                )
                            }
                            Text(id.keyId)
                                .font(.system(size: 12, design: .monospaced))
                                .foregroundStyle(Color.textPrimary)
                                .lineLimit(1)
                            Spacer().frame(height: 8)
                            Text("CREATED")
                                .font(.system(size: 11))
                                .foregroundStyle(Color.textTertiary)
                            Text(String(id.createdAt.prefix(10)))
                                .font(.system(size: 12))
                                .foregroundStyle(Color.textSecondary)
                        }
                        .ssdidCard()
                    }

                    // Pre-rotation status
                    Spacer().frame(height: 4)
                    Text("PRE-ROTATION STATUS")
                        .font(.ssdidCaption)
                        .foregroundStyle(Color.textTertiary)
                        .frame(maxWidth: .infinity, alignment: .leading)

                    VStack(alignment: .leading, spacing: 0) {
                        if hasPreCommitment {
                            HStack(spacing: 8) {
                                Image(systemName: "checkmark")
                                    .font(.system(size: 16, weight: .bold))
                                    .foregroundStyle(Color.success)
                                Text("Next key pre-committed")
                                    .font(.system(size: 14))
                                    .foregroundStyle(Color.success)
                            }
                            if let hash = nextKeyHash {
                                Spacer().frame(height: 8)
                                Text("HASH")
                                    .font(.system(size: 11))
                                    .foregroundStyle(Color.textTertiary)
                                Text(String(hash.prefix(24)) + "...")
                                    .font(.system(size: 12, design: .monospaced))
                                    .foregroundStyle(Color.textSecondary)
                            }
                        } else {
                            HStack(spacing: 8) {
                                Image(systemName: "exclamationmark.triangle")
                                    .font(.system(size: 16))
                                    .foregroundStyle(Color.warning)
                                Text("No pre-commitment")
                                    .font(.system(size: 14))
                                    .foregroundStyle(Color.warning)
                            }
                            Spacer().frame(height: 8)
                            Text("Generate a pre-committed next key for safe rotation.")
                                .font(.system(size: 12))
                                .foregroundStyle(Color.textTertiary)
                        }
                    }
                    .ssdidCard()

                    // Action buttons
                    Spacer().frame(height: 8)

                    if !hasPreCommitment {
                        Button {
                            prepareRotation()
                        } label: {
                            if case .preparing = state {
                                HStack(spacing: 10) {
                                    ProgressView().progressViewStyle(CircularProgressViewStyle(tint: .white))
                                    Text("Preparing...")
                                }
                            } else {
                                Text("Prepare Pre-Commitment")
                            }
                        }
                        .buttonStyle(.ssdidPrimary(enabled: !inProgress))
                        .disabled(inProgress)
                    } else {
                        Button {
                            executeRotation()
                        } label: {
                            if case .rotating = state {
                                HStack(spacing: 10) {
                                    ProgressView().progressViewStyle(CircularProgressViewStyle(tint: Color.bgPrimary))
                                    Text("Rotating...")
                                        .foregroundStyle(Color.bgPrimary)
                                }
                            } else {
                                HStack(spacing: 8) {
                                    Image(systemName: "arrow.triangle.2.circlepath")
                                    Text("Rotate Now")
                                }
                                .foregroundStyle(Color.bgPrimary)
                            }
                        }
                        .font(.ssdidBody.weight(.semibold))
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, 14)
                        .background(!inProgress ? Color.warning : Color.bgElevated)
                        .cornerRadius(12)
                        .disabled(inProgress)
                    }

                    // Warning
                    Text("Current key remains valid for 5 minutes after rotation (grace period).")
                        .font(.system(size: 12))
                        .foregroundStyle(Color.warning)
                        .lineSpacing(4)
                        .padding(14)
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .background(Color.warningDim)
                        .cornerRadius(12)

                    // Status messages
                    if case .success(let message) = state {
                        Text(message)
                            .font(.system(size: 13))
                            .foregroundStyle(Color.success)
                            .padding(14)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .background(Color.successDim)
                            .cornerRadius(12)
                    }
                    if case .error(let message) = state {
                        Text(message)
                            .font(.system(size: 13))
                            .foregroundStyle(Color.danger)
                            .padding(14)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .background(Color.dangerDim)
                            .cornerRadius(12)
                    }

                    // Rotation history
                    if !rotationHistory.isEmpty {
                        Spacer().frame(height: 8)
                        Text("RECENT ROTATIONS")
                            .font(.ssdidCaption)
                            .foregroundStyle(Color.textTertiary)
                            .frame(maxWidth: .infinity, alignment: .leading)

                        ForEach(rotationHistory) { entry in
                            HStack(spacing: 12) {
                                ZStack {
                                    RoundedRectangle(cornerRadius: 10)
                                        .fill(Color.pqcDim)
                                        .frame(width: 36, height: 36)
                                    Image(systemName: "arrow.triangle.2.circlepath")
                                        .font(.system(size: 14))
                                        .foregroundStyle(Color.pqc)
                                }
                                VStack(alignment: .leading, spacing: 2) {
                                    Text("#\(entry.oldKeyFragment) -> #\(entry.newKeyFragment)")
                                        .font(.system(size: 13, design: .monospaced))
                                        .foregroundStyle(Color.textPrimary)
                                    Text(String(entry.timestamp.prefix(10)))
                                        .font(.system(size: 11))
                                        .foregroundStyle(Color.textTertiary)
                                }
                                Spacer()
                            }
                            .padding(14)
                            .background(Color.bgCard)
                            .cornerRadius(12)
                        }
                    }

                    Spacer().frame(height: 20)
                }
                .padding(.horizontal, 20)
            }
        }
        .background(Color.bgPrimary)
        .task {
            identity = await services.vault.getIdentity(keyId: keyId)
            if let id = identity {
                let manager = makeRotationManager(for: id)
                let status = await manager.getRotationStatus(identity: id)
                hasPreCommitment = status.hasPreCommitment
                nextKeyHash = status.nextKeyHash
                rotationHistory = status.rotationHistory.map { entry in
                    RotationHistoryEntry(
                        oldKeyFragment: entry.oldKeyIdFragment,
                        newKeyFragment: entry.newKeyIdFragment,
                        timestamp: entry.timestamp
                    )
                }
            }
        }
    }

    private func makeRotationManager(for identity: Identity) -> KeyRotationManager {
        let provider: CryptoProvider = identity.algorithm.isPostQuantum
            ? services.pqcProvider
            : services.classicalProvider
        return KeyRotationManager(
            vault: services.vault,
            storage: services.storage,
            cryptoProvider: provider,
            keychainManager: services.keychainManager,
            ssdidClient: services.ssdidClient
        )
    }

    private func prepareRotation() {
        guard let id = identity else { return }
        state = .preparing
        Task {
            do {
                let manager = makeRotationManager(for: id)
                let hash = try await manager.prepareRotation(identity: id)
                identity = await services.vault.getIdentity(keyId: keyId)
                hasPreCommitment = true
                nextKeyHash = hash
                state = .success("Pre-commitment created and published to registry")
            } catch {
                state = .error(error.localizedDescription)
            }
        }
    }

    private func executeRotation() {
        guard let id = identity else { return }
        state = .rotating
        Task {
            do {
                let manager = makeRotationManager(for: id)
                let newIdentity = try await manager.executeRotation(identity: id)
                identity = newIdentity
                hasPreCommitment = false
                nextKeyHash = nil
                state = .success("Key rotated successfully. New key: \(newIdentity.keyId)")
            } catch {
                state = .error(error.localizedDescription)
            }
        }
    }
}

File: android/app/src/main/java/my/ssdid/wallet/feature/rotation/KeyRotationScreen.kt
package my.ssdid.wallet.feature.rotation

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import my.ssdid.sdk.domain.model.Identity
import my.ssdid.sdk.domain.rotation.KeyRotationManager
import my.ssdid.sdk.domain.rotation.RotationEntry
import my.ssdid.sdk.domain.rotation.RotationStatus
import my.ssdid.sdk.domain.vault.Vault
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.Warning
import my.ssdid.wallet.ui.theme.*
import javax.inject.Inject

sealed class RotationState {
    object Idle : RotationState()
    object Preparing : RotationState()
    object Rotating : RotationState()
    data class Success(val message: String) : RotationState()
    data class Error(val message: String) : RotationState()
}

@HiltViewModel
class KeyRotationViewModel @Inject constructor(
    private val vault: Vault,
    private val rotationManager: KeyRotationManager,
    savedStateHandle: SavedStateHandle
) : ViewModel() {
    private val keyId: String = savedStateHandle["keyId"] ?: ""

    private val _identity = MutableStateFlow<Identity?>(null)
    val identity = _identity.asStateFlow()

    private val _status = MutableStateFlow<RotationStatus?>(null)
    val status = _status.asStateFlow()

    private val _state = MutableStateFlow<RotationState>(RotationState.Idle)
    val state = _state.asStateFlow()

    init {
        viewModelScope.launch {
            val id = vault.getIdentity(keyId)
            _identity.value = id
            id?.let { _status.value = rotationManager.getRotationStatus(it) }
        }
    }

    fun prepareRotation() {
        val id = _identity.value ?: return
        viewModelScope.launch {
            _state.value = RotationState.Preparing
            rotationManager.prepareRotation(id)
                .onSuccess {
                    _identity.value = vault.getIdentity(keyId)
                    _identity.value?.let { _status.value = rotationManager.getRotationStatus(it) }
                    _state.value = RotationState.Success("Pre-commitment created")
                }
                .onFailure { _state.value = RotationState.Error(it.message ?: "Failed") }
        }
    }

    fun executeRotation() {
        val id = _identity.value ?: return
        viewModelScope.launch {
            _state.value = RotationState.Rotating
            rotationManager.executeRotation(id)
                .onSuccess { newIdentity ->
                    _identity.value = newIdentity
                    _status.value = rotationManager.getRotationStatus(newIdentity)
                    _state.value = RotationState.Success("Key rotated successfully")
                }
                .onFailure { _state.value = RotationState.Error(it.message ?: "Rotation failed") }
        }
    }
}

@Composable
fun KeyRotationScreen(
    onBack: () -> Unit,
    viewModel: KeyRotationViewModel = hiltViewModel()
) {
    val identity by viewModel.identity.collectAsState()
    val status by viewModel.status.collectAsState()
    val state by viewModel.state.collectAsState()

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(BgPrimary)
            .statusBarsPadding().navigationBarsPadding()
    ) {
        // Header
        Row(
            Modifier.padding(start = 8.dp, end = 20.dp, top = 12.dp, bottom = 12.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            IconButton(onClick = onBack) { Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back", tint = TextPrimary) }
            Spacer(Modifier.width(4.dp))
            Text("Key Rotation", style = MaterialTheme.typography.titleLarge)
        }

        LazyColumn(
            Modifier.padding(horizontal = 20.dp),
            verticalArrangement = Arrangement.spacedBy(10.dp)
        ) {
            // Current key info
            item {
                Text("CURRENT KEY", style = MaterialTheme.typography.labelMedium)
                Spacer(Modifier.height(8.dp))
            }

            item {
                identity?.let { id ->
                    Card(
                        Modifier.fillMaxWidth(),
                        shape = RoundedCornerShape(12.dp),
                        colors = CardDefaults.cardColors(containerColor = BgCard)
                    ) {
                        Column(Modifier.padding(16.dp)) {
                            Row(
                                Modifier.fillMaxWidth(),
                                horizontalArrangement = Arrangement.SpaceBetween,
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                Text("KEY ID", fontSize = 11.sp, color = TextTertiary)
                                val algColor = if (id.algorithm.isPostQuantum) Pqc else Classical
                                val algBg = if (id.algorithm.isPostQuantum) PqcDim else ClassicalDim
                                Box(
                                    Modifier
                                        .clip(RoundedCornerShape(4.dp))
                                        .background(algBg)
                                        .padding(horizontal = 8.dp, vertical = 3.dp)
                                ) {
                                    Text(
                                        id.algorithm.name.replace("_", "-"),
                                        fontSize = 10.sp,
                                        fontWeight = FontWeight.SemiBold,
                                        color = algColor
                                    )
                                }
                            }
                            Text(
                                id.keyId,
                                fontSize = 12.sp,
                                color = TextPrimary,
                                fontFamily = FontFamily.Monospace,
                                maxLines = 1
                            )
                            Spacer(Modifier.height(8.dp))
                            Text("CREATED", fontSize = 11.sp, color = TextTertiary)
                            Text(id.createdAt.take(10), fontSize = 12.sp, color = TextSecondary)
                        }
                    }
                }
            }

            // Pre-rotation status
            item {
                Spacer(Modifier.height(4.dp))
                Text("PRE-ROTATION STATUS", style = MaterialTheme.typography.labelMedium)
                Spacer(Modifier.height(8.dp))
            }

            item {
                Card(
                    Modifier.fillMaxWidth(),
                    shape = RoundedCornerShape(12.dp),
                    colors = CardDefaults.cardColors(containerColor = BgCard)
                ) {
                    Column(Modifier.padding(16.dp)) {
                        status?.let { s ->
                            if (s.hasPreCommitment) {
                                Row(verticalAlignment = Alignment.CenterVertically) {
                                    Icon(Icons.Default.Check, contentDescription = "Pre-committed", modifier = Modifier.size(16.dp), tint = Success)
                                    Spacer(Modifier.width(8.dp))
                                    Text("Next key pre-committed", fontSize = 14.sp, color = Success)
                                }
                                s.nextKeyHash?.let { hash ->
                                    Spacer(Modifier.height(8.dp))
                                    Text("HASH", fontSize = 11.sp, color = TextTertiary)
                                    Text(
                                        hash.take(24) + "...",
                                        fontSize = 12.sp,
                                        color = TextSecondary,
                                        fontFamily = FontFamily.Monospace
                                    )
                                }
                            } else {
                                Row(verticalAlignment = Alignment.CenterVertically) {
                                    Icon(Icons.Default.Warning, contentDescription = "Warning", modifier = Modifier.size(16.dp), tint = Warning)
                                    Spacer(Modifier.width(8.dp))
                                    Text("No pre-commitment", fontSize = 14.sp, color = Warning)
                                }
                                Spacer(Modifier.height(8.dp))
                                Text(
                                    "Generate a pre-committed next key for safe rotation.",
                                    fontSize = 12.sp,
                                    color = TextTertiary
                                )
                            }
                        }
                    }
                }
            }

            // Action buttons
            item {
                Spacer(Modifier.height(8.dp))
                val inProgress = state is RotationState.Preparing || state is RotationState.Rotating

                if (status?.hasPreCommitment != true) {
                    Button(
                        onClick = { viewModel.prepareRotation() },
                        modifier = Modifier.fillMaxWidth(),
                        enabled = !inProgress,
                        shape = RoundedCornerShape(12.dp),
                        colors = ButtonDefaults.buttonColors(containerColor = Accent)
                    ) {
                        if (state is RotationState.Preparing) {
                            CircularProgressIndicator(Modifier.size(20.dp), color = BgPrimary, strokeWidth = 2.dp)
                            Spacer(Modifier.width(10.dp))
                        }
                        Text("Prepare Pre-Commitment", fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
                    }
                } else {
                    Button(
                        onClick = { viewModel.executeRotation() },
                        modifier = Modifier.fillMaxWidth(),
                        enabled = !inProgress,
                        shape = RoundedCornerShape(12.dp),
                        colors = ButtonDefaults.buttonColors(containerColor = Warning)
                    ) {
                        if (state is RotationState.Rotating) {
                            CircularProgressIndicator(Modifier.size(20.dp), color = BgPrimary, strokeWidth = 2.dp)
                            Spacer(Modifier.width(10.dp))
                        }
                        Icon(Icons.Default.Refresh, contentDescription = null, modifier = Modifier.size(18.dp))
                        Spacer(Modifier.width(8.dp))
                        Text("Rotate Now", fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
                    }
                }
            }

            // Warning
            item {
                Card(
                    Modifier.fillMaxWidth(),
                    shape = RoundedCornerShape(12.dp),
                    colors = CardDefaults.cardColors(containerColor = WarningDim)
                ) {
                    Text(
                        "Current key remains valid for 5 minutes after rotation (grace period).",
                        modifier = Modifier.padding(14.dp),
                        fontSize = 12.sp,
                        color = Warning,
                        lineHeight = 18.sp
                    )
                }
            }

            // Status messages
            if (state is RotationState.Success) {
                item {
                    Card(
                        Modifier.fillMaxWidth(),
                        shape = RoundedCornerShape(12.dp),
                        colors = CardDefaults.cardColors(containerColor = SuccessDim)
                    ) {
                        Text(
                            (state as RotationState.Success).message,
                            modifier = Modifier.padding(14.dp),
                            fontSize = 13.sp,
                            color = Success
                        )
                    }
                }
            }
            if (state is RotationState.Error) {
                item {
                    Card(
                        Modifier.fillMaxWidth(),
                        shape = RoundedCornerShape(12.dp),
                        colors = CardDefaults.cardColors(containerColor = DangerDim)
                    ) {
                        Text(
                            (state as RotationState.Error).message,
                            modifier = Modifier.padding(14.dp),
                            fontSize = 13.sp,
                            color = Danger
                        )
                    }
                }
            }

            // Rotation history
            val history = status?.rotationHistory ?: emptyList()
            if (history.isNotEmpty()) {
                item {
                    Spacer(Modifier.height(8.dp))
                    Text("RECENT ROTATIONS", style = MaterialTheme.typography.labelMedium)
                    Spacer(Modifier.height(8.dp))
                }
                items(history) { entry ->
                    Card(
                        Modifier.fillMaxWidth(),
                        shape = RoundedCornerShape(12.dp),
                        colors = CardDefaults.cardColors(containerColor = BgCard)
                    ) {
                        Row(Modifier.padding(14.dp), verticalAlignment = Alignment.CenterVertically) {
                            Box(
                                Modifier
                                    .size(36.dp)
                                    .clip(RoundedCornerShape(10.dp))
                                    .background(PqcDim),
                                contentAlignment = Alignment.Center
                            ) {
                                Icon(Icons.Default.Refresh, contentDescription = "Rotated", modifier = Modifier.size(14.dp), tint = Pqc)
                            }
                            Spacer(Modifier.width(12.dp))
                            Column(Modifier.weight(1f)) {
                                Text(
                                    "#${entry.oldKeyIdFragment} \u2192 #${entry.newKeyIdFragment}",
                                    fontSize = 13.sp,
                                    fontFamily = FontFamily.Monospace,
                                    color = TextPrimary
                                )
                                Text(
                                    entry.timestamp.take(10),
                                    fontSize = 11.sp,
                                    color = TextTertiary
                                )
                            }
                        }
                    }
                }
            }

            item { Spacer(Modifier.height(20.dp)) }
        }
    }
}
